### ISSUE TITLE
The training step revalidates and carries what the model never reads

### ISSUE THREAD

**Issue by mmmfrieddough** [mmmfrieddough/anthro-chess#197]
## Goal

Delete per-step work that produces nothing the model reads. Every item below is
a removal; none of them should be replaced with a reimplementation.

## Findings

Measured on one RTX 4090 at batch 16 by 150 plies.

**1. `MoveModelBatch.validate()` runs twice per micro-batch.** Every factory
(`from_sequence_batch`, `from_decision_contexts`, `stack`) already calls it, and
`encode_history` calls it again on every forward pass. There are zero direct
`MoveModelBatch(...)` constructions anywhere in `src` or `tests`, so the
forward-pass call re-validates an object that is validated by construction.
About 0.5 ms of host work per call.

**2. The causal mask is rebuilt in Python every batch.** `collate_sequences`
builds a T-by-T nested tuple of bools — 22,500 comparisons at T=150 — which is
then pickled through the streaming loader, converted to a device tensor, and
handed to `nn.TransformerEncoder` as an explicit mask. `is_causal=True` says the
same thing and needs none of it.

**3. `src_key_padding_mask` is redundant.** Padding is right-aligned, so a real
query at position `i` attends only to `j <= i`, all of which are real. Padded
query rows produce values that `forward`'s `masked_fill` and the loss mask
already discard. Dropping it also removes the all-padding-row NaN hazard that
key padding masks are known for.

**4. `legal_action_ids` rides along in every training batch** and is read only
by `_reject_illegal_active_targets`. The model never touches it. It is about 26%
of the streaming loader's pickle payload: dropping it takes the batch from
0.58 MB to 0.43 MB and the parent's `pickle.loads` from 2.60 ms to 2.01 ms.
Evaluation's legality metrics do use it, so it needs to stop travelling with
training batches rather than disappear.

## Why now

`docs/planning/cuda-training-proof.md` found the step host-bound, with the card
at roughly 11% utilization. All four items are host work.

Items 2 and 3 also unlock the flash and cuDNN attention backends, which reject a
non-null `attn_mask`. That is worth nothing at `model_dim = 64` and a great deal
after #54:

| | explicit mask | `is_causal` |
| --- | ---: | ---: |
| `model_dim = 64` | 0.047 ms | 0.029 ms |
| `model_dim = 512` | 0.083 ms | 0.061 ms |

Items 1 and 4 are the ragged Python structures the same document names as the
cause of the `torch.compile` recompile-limit failure.

## Acceptance Criteria

- The forward path does not revalidate a batch produced by a factory, and the
  validation contract is stated at the construction boundary instead.
- Attention uses the default causal path. The explicit causal mask and the key
  padding mask are gone from the model, the batch, and the collate step.
- `legal_action_ids` no longer travels in the training batch; evaluation
  legality paths keep what they need.
- A before-and-after CUDA step-time reading is reported.
- CPU suite green. No new configuration dial.

## Notes

Sequenced ahead of the array-representation issue on purpose: both touch
`models/batching.py` and `data/loading.py`, and deleting first shrinks what that
change has to convert.

## References

- `docs/planning/cuda-training-proof.md`
- `docs/training-and-runtime.md`, `docs/architecture.md`

**Comment by mmmfrieddough** [comment]
## Four more, from a full pass over the invocation-to-Torch path

All measured at batch 16 by 150 plies on one RTX 4090, against a 5.44 ms
forward and backward. Each replaces something hand-written with a Torch default
or removes it outright; none adds a dial.

**1. `forward` masks padded logits to zero, and nothing reads them.**
`logits.masked_fill(~attention_mask.unsqueeze(-1), 0.0)` writes 4.7M elements per
forward and carries them through the backward pass. Every consumer was checked:
`inference.runner.predict_batch` indexes each row at its own real ply,
`action_logits` returns the raw tensor, `evaluation.validation` gates on the
loss mask, and `score_positions` gates on `_active_batch`. **0.150 ms**, pure
deletion.

**2. The sinusoidal position table is rebuilt on every forward.**
`_sinusoidal_positions` recomputes `arange`, `exp`, `sin`, `cos` and two strided
scatters for a value that depends only on ply index and model width. As a
registered buffer indexed by `ply_indices` it is one gather: **0.081 ms to
0.007 ms**. This one adds a buffer rather than deleting code, which is worth
saying plainly, but it removes the per-step arithmetic and about four launches.

**3. The masked loss uses boolean indexing where `ignore_index` is the default
idiom.** `F.cross_entropy(logits[loss_mask], targets[loss_mask])` gathers into a
**data-dependent shape**, which is the second source of the dynamic shapes that
`docs/planning/cuda-training-proof.md` blames for the `torch.compile` recompile
limit — the first being `legal_action_ids`. Emitting `-100` for masked targets
from the loader and calling `F.cross_entropy(..., ignore_index=-100)` is
numerically identical (verified) and static-shaped: **0.356 ms to 0.142 ms**. It
also retires the `torch.any(loss_mask)` guard, since no target enabled means no
contribution rather than an error.

**4. The timing fields are packed, pickled, and then rejected.**
`collate_sequences` builds `time_initial_ms`, `time_increment_ms`,
`player_clock_ms`, and `opponent_clock_ms` as four value-and-presence pairs,
sends all eight arrays through the loader's IPC, and `_validated_plies` raises if
any of them is populated. The move-time head is Milestone 5 work. Until then
these are carried the length of the pipeline to be discarded at the model
boundary.

## Also in scope, same theme

The model performs per-forward fixups that are properties of the encoding rather
than of the model: the en-passant token's `+1` offset, the previous-action fill
value, and `log1p` on the two rule counters. Each is a deterministic function of
loader output, so each belongs in the encoding that `#198` is converting to
arrays, computed once rather than per step.

`optimizer.zero_grad(set_to_none=True)` states the Torch default; the argument
can go.

## Total

Together with the duplicate `validate()` already in this issue, roughly
**1.2 ms off a 5.44 ms compute step, about 23%**, entirely from deleting work or
using a Torch default. Item 3 is the one worth doing even if it were free,
because it is half of what stands between this model and `torch.compile`.

**Comment by mmmfrieddough** [comment]
## Correction: `is_causal=True` is not a drop-in, and flash needs bfloat16

Two claims in the earlier comment on the attention change are wrong. Testing
them on Torch 2.13.0+cu130 before implementing, rather than after.

**`is_causal=True` alone does not apply causal masking.** Passing it to
`nn.TransformerEncoder` with no `mask` runs without error and returns a
*different* result from the masked version:

```
enc(x, mask=~causal)                  # reference
enc(x, is_causal=True)                # runs, allclose -> False
enc(x, mask=~causal, is_causal=True)  # allclose -> True
```

`is_causal` behaves as a hint alongside the mask rather than as a replacement
for it. Dropping the mask and setting the flag would silently produce a
non-causal model that trains and evaluates without complaint — a leakage bug
with no failing test. **The mask must still be passed.** Any implementation of
this item owes an equivalence check against the current masked output, not an
assumption.

**Flash attention is unavailable in float32 regardless of masking.** The earlier
comment claimed `is_causal` unlocks the flash and cuDNN backends. Measured, with
fp32 inputs:

```
SDPA explicit bool mask   flash available: False
SDPA is_causal=True       flash available: False
```

Flash requires half precision. The current default is `precision = "float32"`,
so flash is off either way and this change unlocks nothing on its own. The
0.047 ms to 0.029 ms difference measured earlier is real but comes from not
materializing and reading a T-by-T mask, not from a better backend.

Getting flash is a *bfloat16 plus causal* combination, which makes it a question
for `#200` and the capacity work rather than a benefit of this issue. It also
gives `precision = "bfloat16-mixed"` a second reason to exist beyond the memory
argument it was retained on.

**What survives:** the key-padding mask is still redundant under right-aligned
padding and still worth removing, and not materializing the causal mask per
batch is still worth doing. The framing changes from "unlock flash" to "stop
building and reading a tensor that encodes a property the API can express."

**One more, relevant to `#200` rather than here.** `nn.TransformerEncoder` with
any explicit `mask=` fails `torch.compile(fullgraph=True)` on this build with
`Could not guard on data-dependent expression Eq(u0, 1)`. `nn.MultiheadAttention`
itself is fine in every configuration tested, including with both masks — the
problem is the encoder wrapper's mask handling, not attention. An equivalent
model written against `F.scaled_dot_product_attention` directly compiles to one
graph with zero breaks and takes CUDA graphs. That is a bigger architectural
question than this issue, and it is recorded on `#200`.

**Comment by mmmfrieddough** [comment]
## Maintainer decision: the timing fields stay

Item 4 of the earlier comment — removing `time_initial_ms`,
`time_increment_ms`, `player_clock_ms`, and `opponent_clock_ms` from the packing
and loader path — **is out of scope for this issue**. They stay as they are.

The observation stands that they are packed, sent through the loader's IPC, and
then rejected at the model boundary. It is not worth acting on ahead of the
timing work they exist for, and the tracker already places timing conditioning in
this milestone rather than a later one.

The other three items in that comment are unaffected: the padded-logit
`masked_fill`, the per-forward sinusoidal table, and the boolean-indexed loss.

**Comment by mmmfrieddough** [comment]
Claimed for implementation by Claude Code (Opus 5).

**Comment by mmmfrieddough** [comment]
Implemented in #206. Two notes on scope that belong here rather than only in the pull request.

**The three "also in scope, same theme" items are deferred to #198.** The en-passant `+1` offset, the previous-action fill value, and `log1p` on the two rule counters are all still computed per forward pass. The comment places them "in the encoding that #198 is converting to arrays", and doing them now means choosing a representation that #198 immediately rewrites — both issues already touch `models/batching.py` and `data/loading.py`, which is the reason this one was sequenced first. Two of the three also change what `encoding_identity()` describes, which is part of checkpoint compatibility, so they are a larger decision than a deletion. Everything else in the issue and the surviving comment items are in the pull request.

**`F.cross_entropy` with every target ignored returns NaN, not zero.** The comment retires the `torch.any(loss_mask)` guard on the grounds that "no target enabled means no contribution rather than an error". Measured on Torch 2.13.0+cu130, it is a non-finite loss instead. The guard is still gone, because it is a device read on every micro-batch and the loader cannot produce an empty mask, and a run already refuses to continue on a non-finite loss — but the failure mode is worth having written down rather than found later. There is a test pinning it.

The reading, taken on one RTX 4090 against `lichess-blitz-proof-30k-v4` at the settings `docs/planning/cuda-training-proof.md` used: profiled step 13.79 ms to 11.27 ms, with forward and backward 6.48 ms to 5.05 ms. That is 22.1% off the compute step against the 23% this issue predicted. Real-timestep logits and the loss are bit-identical to the previous path on both CPU and CUDA.

**Comment by mmmfrieddough** [comment]
Two findings from the self review that are out of scope here and want their own issues, both measured on one RTX 4090 rather than reasoned.

**`encode_game` reconstructs legal actions for every ply regardless of who asked.** This pull request stops them travelling in a training batch, which is what the issue asked for, but the reconstruction upstream is untouched and is far larger. `legal_action_ids()` is **65% of `encode_game`** — 3,946 µs of 6,053 µs on an 80-ply game at 31.2 legal actions per ply. A 16-game training batch therefore burns roughly **63 ms of worker CPU** building lists nobody reads, against the ~0.9 ms of pickling the flag saves. A per-ply `board.is_legal(decode_move(...))` covering the same target check costs 709 µs, 18% of it. Not a one-line swap: `legal_action_ids` also decides resignation and draw-claim legality, so the terminal-action cases need handling separately. This is the biggest remaining item in the loader path and it is bigger than everything this issue removed.

**`validate()` is now the largest host cost left in the step.** 282–289 µs per micro-batch on CUDA at batch 16, essentially all of it `_reject_invalid_values` — about twenty queued reductions plus one `.cpu().tolist()` read — and flat in sequence length. That is ~7% of the 4.2 ms step. This issue deleted the duplicate call; the surviving one is in `from_sequence_batch` and runs every micro-batch on batches whose shape and provenance do not vary across steps. Recording the residue rather than proposing a mechanism.

Also worth correcting against the correction comment: on the locked build (Torch 2.13.0+cu130) `nn.TransformerEncoder` **raises** `Need attn_mask if specifying the is_causal hint` when given the flag with no mask, in both train and eval mode — it does not run and return a different result. The conclusion the comment drew is unchanged and the mask is still passed, but the silent-leakage-with-no-failing-test hazard cannot land on this build.

**Comment by mmmfrieddough** [comment]
Correction on where the two findings above live. This issue closes when #206 merges, so a comment here is a record rather than a queue — neither would have been picked up again. Both now have their own issues, attached to the milestone 4 tracker and blocked by #198, which rewrites the surfaces they both touch:

- #221 — the encoding reconstructs legal actions for every ply regardless of who reads them
- #222 — the batch is revalidated on every micro-batch of every step

The measurements stay above; the actionable form is there.
